src/proj_equirect.jl

```diff
@@ -242,41 +231,109 @@ end
 # block-diagonal operator
 # ================================================
 
+
 function BlockDiagEquiRect{B}(block_matrix::A, proj::P) where {B<:Basis, P<:ProjEquiRect, T, A<:AbstractArray{T}}
-    BlockDiagEquiRect{B,P,T,A}(block_matrix, Ref{A}(), proj)
+    BlockDiagEquiRect{B,P,T,A}(block_matrix, proj)
 end
 
-size(L::BlockDiagEquiRect) = (fill(L.proj.Nx * L.proj.Ny, 2)...,)
-
-function sqrt(L::BlockDiagEquiRect{B}) where {B}
-    if !isassigned(L.blocks_sqrt)
-        L.blocks_sqrt[] = mapslices(sqrt, L.blocks, dims=(1,2))
+# Allows construction by a vector of blocks
+function BlockDiagEquiRect{B}(vector_of_blocks::Vector{A}, proj::P) where {B<:Basis, P<:ProjEquiRect, T, A<:AbstractMatrix{T}}
+    block_matrix = Array{T}(undef, size(vector_of_blocks[1])..., length(vector_of_blocks))
+    for b in eachindex(vector_of_blocks)
+        block_matrix[:,:,b] .= vector_of_blocks[b]
     end
-    BlockDiagEquiRect{B}(L.blocks_sqrt[], L.proj)
+    BlockDiagEquiRect{B}(block_matrix, proj)
 end
 
-*(L::BlockDiagEquiRect{B}, f::EquiRectField) where {B<:Basis} = L * B(f)
 
-function *(B::BlockDiagEquiRect{AzFourier}, f::EquiRectAzFourier)
-    promote_metadata_strict(B.proj, f.proj) # ensure same projection
-    EquiRectAzFourier(@tullio(Bf[p,iₘ] := B.blocks[p,q,iₘ] * f.arr[q,iₘ]), f.metadata)
+# ## make BlockDiagEquiRect an iterable over the last index
+# ... so that 
+#     `M½ = BlockDiagEquiRect{AzFourier}(sqrt.(Hermitian.(M)), M.proj)`
+# works
```

Comment:
If I can leave a quick comment here just glancing at this, but I don't think we should do this. `M` is supposed to implicity represent an `Npix-by-Npix` matrix, and so `sqrt.(M)` is supposed to mean element-wise sqrt of that matrix, so we should not override that with a different meaning. From a practical point, I'm fairly certain this code will not work as-is on GPU. 

Was there a reason to ditch  https://github.com/marius311/CMBLensing.jl/pull/64/commits/a3b3624ce6a14ac90865d63b0bc42368de58421b#diff-99ba5f70d2ac8705e7c12fd910135566888ca2761d42cae62e905ce5f20c150fR91-R96 ? That does keep the linear algebra meaning of everything, and fwiw also already works on GPU.

Reply:
> M is supposed to implicity represent an Npix-by-Npix matrix
Alright.  We had a different perspective. I was coming from the viewpoint that `M` _does_ represent the list of matrix blocks (that happens to know how to operate) rather than the full matrix itself. From that perspective stuff like this seems natural 

```
Precon⁻¹ = map(EB, Beam, Noise) do S, B, N
       pinv(B * S * B' + N)
end |> BlockDiagEquiRect

phi_sim = map(Phi▪, white_noise_field) do Σ,v
      Matrix(cholesky(Σ).L)*v
end |> AzFourier 
```

and once you have that then its tempting to want `Precon⁻¹ =  BlockDiagEquiRect(@. pinv(Beam * EB * Beam' + Noise))`. 

Perhaps the way to stay within your viewpoint is to simply overload `mapslices` or something?

Reply:
I see. Yea, the guiding principle for CMBLensing is to write linear algebra expressions with no assumptions at all about what types or projections the fields or operators are in. So if you have

```julia
EB :: BlockDiagEquiRect
Beam :: BlockDiagEquiRect
Noise :: BlockDiagEquiRect
```

then the code

```julia
Beam * EB * Beam' + Noise
```

should just work (and its the exact code that would also work if those were e.g. `Diagonal{<:FlatFourier}`, or `FlatIEBCov`, or etc...). 

That's actually easy, we just need to define:

```julia
*(A::BlockDiagEquiRect, B::BlockDiagEquiRect) = BlockDiagEquiRect(mapslices(*, A, B, dims=(1,2)))
+(A::BlockDiagEquiRect, B::BlockDiagEquiRect) = BlockDiagEquiRect(mapslices(+, A, B, dims=(1,2)))
adjoint(A::BlockDiagEquiRect) = BlockDiagEquiRect(mapslices(adjoint, )A, dims=(1,2))
# ^slight pseudo-code where I'm ignoring the B type parameter and proj
```

Similarly, instead of 

```julia
phi_sim = map(Phi▪, white_noise_field) do Σ,v
      Matrix(cholesky(Σ).L)*v
end |> AzFourier 
```

its just `sqrt(L) * white_noise_field`. The implementation I had before [here](https://github.com/marius311/CMBLensing.jl/pull/64/commits/a3b3624ce6a14ac90865d63b0bc42368de58421b#diff-99ba5f70d2ac8705e7c12fd910135566888ca2761d42cae62e905ce5f20c150fR91-R96) I don't think needs any changing for this one to work. 

I did have the sqrt of the operator lazily stored inside the `BlockDiagRectOp`, so that its only calculated once the first time you call `sqrt(L)` for a given operator. The reason it was stored in there is because for most operators, it will come up repeatedly during sampling and we don't want to recompute the `sqrt` each time, and this I thought was a nice way to store the `sqrt` without having to ever think about it. However, since you wont need it for *all* operators, (eg not for the beam), you dont want to always store it, hence the "laziness". We might have to do a similar for `pinv` which also comes up alot, but my hope is that thats fast enough it can just be done on the fly and we can save memory. 

I think its totally fine to have this mapping functionality you have above for convenience though, I would just call it `mapblocks` or something, since I don't think it fits into `mapslices` either.

Reply:
The only tricky thing about this LinearAlgebra purity is that it seems like we should then add stuff like Hermitian/Symmetric/LowerTriangular/ ... etc, which would be a pain. 

For example, when appropriate I usually apply Hermitian/Symmetric before passing to `cholesky` or `sqrt`. Also sometimes it is convenient to store `cholesky(A).L` if I need both the `sqrt` and its inverse but memory is tight. When using something like `mapblocks` the mapped function utilizes all this internally, then once done, can re-map to blocks of matrices. 

To make this comment more specific for `sqrt`... I'm not sure if I should just assume it will only be applied to Hermitian/Symmetric and then internally apply `sqrt∘Hermitian` or `sqrt∘Symmetric` depending on the eltype? Doesn't seem right from the LinearAlgebra perspective.  Note, things like Beams are not going to be symmetric (but then again we will never be `sqrt`ing them). 

For now, I'll start by implementing just the basics `*, +, adjoint` and some basic `mapblocks` methods. These will give us the functionality we need, then can work on streamlining things like `sqrt` and `pinv` later.

Reply:
I forgot to mention another advantage of something like `mapblocks`, if we ever get to a situation where memory is really tight, we can sometimes store two symmetric block operators into one `BlockDiagEquiRect`. Upper triangular holds one, lower triangular holds the other (in the complex case the diagonal will be real so we can store them separately as real/imag). Using `mapblocks` will allow us to extract and multiply whichever we want on the fly. 

I would only utilize this in a pinch, but I like being able to have it around for a plan B.

Reply:
> For now, I'll start by implementing just the basics *, +, adjoint and some basic mapblocks methods.

Yes, this is the right approach. We don't have to implement everything, just the stuff we need. I think having `mapblocks` is good and is going to make writing most those including `sqrt` and `pinv` super simple one-liners. 

> I'm not sure if I should just assume it will only be applied to Hermitian/Symmetric and then internally apply sqrt∘Hermitian or sqrt∘Symmetric depending on the eltype?

I would say lets not assume `BlockDiagEquiRect` necessarily represents something symmetric, b/c 1) `Symmetric`/`Hermitian` wrappers dont work on GPU anyway 2) on CPU its essentially a performance optimization, and I think at this point its premature optimization (especially since we lazily store the sqrt). So lets just code it as if its allowed to be asymetric, hence `sqrt(L) = mapblocks(sqrt, L)` rather than `mapblocks(sqrt∘Hermitian, L)` 


> Note, things like Beams are not going to be symmetric

Which types of operators give you a block-diagonal but _asymmetric_ operator? Do we expect to have any? I haven't thought about this before. A rotationally symmetric beam (like we have) should be symmetric I believe.

Reply:
> A rotationally symmetric beam (like we have) should be symmetric I believe.
The beam operators will look like `B * Diagonal(Ω)` where `Ω` represents the pixel area along the polar coordinate (and `B` the symmetric operator representing the rotationally symmetric beam). Basically ` Diagonal(Ω)` corresponds to the `dS` in the convolution formula. So unless something magic happens `B * Diagonal(Ω)` shouldn't be symmetric.

Reply:
I need to think about this more but may not have time for a few days. Just to say now, but its confusing to me how a beam operator, which corresponds to convolution by a rotationally symmetric beam profile, could be an asymmetric operator, while the `Cf` operator which corresponds to convolution by the rotationally symmetric CMB real-space profile would be a symmetric opertaor. I'm guessing the reason has to do with how you think about `Cf` as a covariance (hence its obvoiusly symmetric) vs. a convolution. And a bit further, it probably has to do with where you decide to stick in the factor of `Ω`, which has implications for what `white_noise`, `dot`, `logdet`, and ultimately the AD gradient rules etc... give. In any case this will surely get ironed out as we get all those remaining tests passing.

Reply:
>  I'm guessing the reason has to do with how you think about Cf as a covariance (hence its obvoiusly symmetric) vs. a convolution. And a bit further, it probably has to do with where you decide to stick in the factor of Ω

Thats exactly it! I think its a bit tricky to see when working in the flat sky since the grid area element is a constant (and therefore just an implicit scalar multiplication which can be put different places depending on if you want to use an inner product that mimics an integral or is just a straight sum). For `BlockDiagEquiRect` the inner product is just a sum (i.e. standard vector dot),  and the linear operators operate by simple matrix mult. 

The beam looks like `symmetricB * Diagonal(Ω) * f ≈ ∫ B(⋅,x) f(x) dΩ` partly due to the fact that I'm constructing `symmetricB` via the matrix product viewpoint and not the integral operator viewpoint. 

Another way to see why the covariance operator and the beam operator need to be different in the matrix product viewpoint is to notice that the constraints are different. For unit variance covariance matrices we need `diag(Σ) .== 1` but for beams need something like `sum(B[i,:]) .== 1` for each row i so it behaves like an averaging operator.

Reply:
What about field dot products? Since we have `dot(f, g) = ∫ g(x) f(x) dΩ` shouldn't this be `sum(f[i] * g[i] * Ω[i])`, so not just the vector dot? Is that consistent with `BlockDiagEquiRect` being the vector dot like you said above?